megamek/src/megamek/utilities/ai/DatasetParser.java

```diff
@@ -0,0 +1,196 @@
+                    while ((line = reader.readLine()) != null && (!line.startsWith("PLAYER_ID\tENTITY_ID"))) {
+                        if (line.trim().isEmpty()) continue;
+                        if (line.startsWith("ROUND")) break;
+                        UnitState state = parseStateLine(line);
+                        if (currentRound == null) {
+                            currentRound = state.round();
+                        } else if (currentRound != state.round()) {
+                            throw new RuntimeException("State block has inconsistent rounds");
+                        }
+                        states.add(state);
+                    }
+
+                    if (currentRound == null) {
+                        throw new RuntimeException("State block has no valid states");
+                    }
+                    actionAndStates.add(new ActionAndState(currentRound, action, states));
+
+                    // If line is an action header, the outer loop will handle it
+                    // we are not currently parsing and using attack actions on the dataset right now.
+                }
+            }
+        } catch (IOException e) {
+            throw new RuntimeException("Error reading file", e);
+        }
+
+        idOffset = highestEntityId + 1;
+        return this;
+    }
+
+    /**
+     * Get the training dataset from the parsed files.
+     * This will throw an exception if no actions and states were found or no parsing was done.
+     * @return The training dataset
+     */
+    public TrainingDataset getTrainingDataset() {
+        if (actionAndStates.isEmpty()) {
+            throw new RuntimeException("No actions and states found in dataset, you need to parse a file first");
+        }
+        return new TrainingDataset(actionAndStates);
+    }
+
+    private UnitAction parseActionLine(String actionLine) {
+        try {
+            String[] parts = actionLine.split("\t");
+            int entityId = Integer.parseInt(parts[1]) + idOffset;
+            int facing = Integer.parseInt(parts[4]);
+            int fromX = Integer.parseInt(parts[5]);
+            int fromY = Integer.parseInt(parts[6]);
+            int toX = Integer.parseInt(parts[7]);
+            int toY = Integer.parseInt(parts[8]);
+            int hexesMoved = Integer.parseInt(parts[9]);
+            int distance = Integer.parseInt(parts[10]);
+            int mpUsed = Integer.parseInt(parts[11]);
+            int maxMp = Integer.parseInt(parts[12]);
+            double mpP = Double.parseDouble(parts[13]);
+            double heatP = Double.parseDouble(parts[14]);
+            double armorP = Double.parseDouble(parts[15]);
+            double internalP = Double.parseDouble(parts[16]);
+            boolean jumping = parts[17].equals("1");
+            boolean prone = parts[18].equals("1");
+            boolean legal = parts[19].equals("1");
+            // Need to check if the unit in the previous turn had moved or not
```

Comment:
I highly, highly recommend replacing all of these indexes with constants or enum values or something.

Reply:
Ok, ok... I will make a proper parser.